crates/turbo-tasks-memory/src/gc.rs

```diff
@@ -1,224 +1,331 @@
-        total_compute_duration: u8,
-    },
-    /// Unload cells that are currently not read by any task. This might cause
-    /// the task to recompute when these cells are read.
-    EmptyUnusedCells {
-        compute_duration: SmallDuration<10_000>,
-    },
-    /// Unload all cells, and continue tracking them valueless. This might cause
-    /// the task and dependent tasks to recompute when these cells are read.
-    EmptyCells {
-        /// Aggregated recompute time. Stored as 2^x milliseconds to bucket
-        /// tasks and avoid frequent revalidation.
-        total_compute_duration: u8,
-        /// The age of the task. Stored as 2^x seconds to
-        /// bucket tasks and avoid frequent revalidation.
-        age: Reverse<u8>,
-    },
-    Placeholder,
-}
-
-/// Statistics about actions performed during garbage collection.
-#[derive(Default, Debug)]
-pub struct GcStats {
-    /// How many tasks were unloaded.
-    pub unloaded: usize,
-    /// How many unused cells were emptied.
-    pub empty_unused: usize,
-    /// How many unused cells were emptied (on the fast path).
-    pub empty_unused_fast: usize,
-    /// How many used cells were emptied.
-    pub empty_cells: usize,
-    /// How often the priority of a task was updated.
-    pub priority_updated: usize,
-    /// How often the priority of a task was updated (on the fast path).
-    pub priority_updated_fast: usize,
-    /// How many tasks were checked but did not need to have any action taken.
-    pub no_gc_needed: usize,
-    /// How many tasks were checked but were in a state where no action could be
-    /// taken.
-    pub no_gc_possible: usize,
+#[derive(Debug, Default, PartialEq, Eq, PartialOrd, Ord, Copy, Clone)]
+pub struct GcPriority {
+    // Memory usage devided by compute duration. Specifies how efficient garbage collection would
+    // be with this task. Higher memory usage and lower compute duration makes it more likely to be
+    // garbage collected.
+    memory_per_time: u16,
 }
 
 /// State about garbage collection for a task.
-#[derive(Debug, Default)]
+#[derive(Clone, Copy, Debug, Default)]
 pub struct GcTaskState {
-    pub inactive: bool,
+    /// Memory usage divided by compute duration. Specifies how efficient
+    /// garbage collection would be with this task. Higher memory usage and
+    /// lower compute duration makes it more likely to be garbage collected.
+    pub memory_per_time: u16,
+    /// The generation where the task was last accessed.
+    pub generation: u32,
+}
+
+impl GcTaskState {
+    pub fn priority(&self) -> GcPriority {
+        GcPriority {
+            memory_per_time: self.memory_per_time,
+        }
+    }
+
+    pub(crate) fn execution_completed(
+        &mut self,
+        duration: Duration,
+        memory_usage: usize,
+        generation: u32,
+    ) {
+        self.generation = generation;
+        self.memory_per_time = ((memory_usage + TASK_BASE_MEMORY_USAGE) as u64
```

Comment:
Not something that needs to change, but a couple observation:

- This seems like it could also be stored as an "order of magnitude" type value, similar to `ExpMillisDuration`. That would decrease the risk of accidental saturation of the `u16` type (potentially due to a bad selection of units), and might mean we could also use a `u8` instead of a `u16` (though admittedly you're not actually saving anything after memory alignment).

- If you wanted, you could probably just store this as a u32 with no extra cost, assuming `GcTaskState` will be aligned to 64 bits, though that's admittedly a fragile assumption (e.g. if you want to add more fields later).

Reply:
It only run once into the u16 limit for a task that allocated 134MB in 43µs^^